src/xwayland/xwm/surface.rs

```diff
@@ -0,0 +1,878 @@
+                x if *x == self.atoms._NET_WM_WINDOW_TYPE_TOOLTIP => Some(WmWindowType::Tooltip),
+                x if *x == self.atoms._NET_WM_WINDOW_TYPE_UTILITY => Some(WmWindowType::Utility),
+                _ => None,
+            })
+    }
+
+    fn change_net_state(&self, added: &[Atom], removed: &[Atom]) -> Result<(), ConnectionError> {
+        let conn = self.conn.upgrade().ok_or(ConnectionError::UnknownError)?;
+        conn.grab_server()?;
+
+        let props = conn
+            .get_property(
+                false,
+                self.window,
+                self.atoms._NET_WM_STATE,
+                AtomEnum::ATOM,
+                0,
+                1024,
+            )?
+            .reply_unchecked()?;
+        let mut new_props = props
+            .and_then(|props| Some(props.value32()?.collect::<HashSet<_>>()))
+            .unwrap_or_default();
+        new_props.retain(|p| !removed.contains(p));
+        new_props.extend(added);
+        let new_props = Vec::from_iter(new_props.into_iter());
+
+        conn.change_property32(
+            PropMode::REPLACE,
+            self.window,
+            self.atoms._NET_WM_STATE,
+            AtomEnum::ATOM,
+            &new_props,
+        )?;
+        self.update_net_state()?;
+        conn.ungrab_server()?;
+        conn.flush()?;
+        Ok(())
+    }
+
+    fn input_mode(&self) -> InputMode {
+        let state = self.state.lock().unwrap();
+        match (
+            state.hints.as_ref().and_then(|hints| hints.input).unwrap_or(true),
+            state.protocols.contains(&WMProtocol::TakeFocus),
+        ) {
+            (false, false) => InputMode::None,
+            (true, false) => InputMode::Passive, // the default
+            (true, true) => InputMode::LocallyActive,
+            (false, true) => InputMode::GloballyActive,
+        }
+    }
+
+    pub(super) fn update_properties(&self, atom: Option<Atom>) -> Result<(), ConnectionError> {
+        match atom {
+            Some(atom) if atom == self.atoms._NET_WM_NAME || atom == AtomEnum::WM_NAME.into() => {
+                self.update_title()
+            }
+            Some(atom) if atom == AtomEnum::WM_CLASS.into() => self.update_class(),
+            Some(atom) if atom == self.atoms.WM_PROTOCOLS => self.update_protocols(),
+            Some(atom) if atom == self.atoms.WM_HINTS => self.update_hints(),
+            Some(atom) if atom == AtomEnum::WM_NORMAL_HINTS.into() => self.update_normal_hints(),
+            Some(atom) if atom == AtomEnum::WM_TRANSIENT_FOR.into() => self.update_transient_for(),
+            Some(_) => Ok(()), // unknown
+            None => {
+                self.update_title()?;
+                self.update_class()?;
+                self.update_protocols()?;
+                self.update_hints()?;
+                self.update_normal_hints()?;
+                self.update_transient_for()?;
+                self.update_net_state()?;
```

Comment:
Perhaps a comment saying that `_NET_WM_STATE` may only be modified by the WM and thus does not need to be in the list above?